CLM Fortran compile error

Ethan Coon reported the following issue with CLM:

I’m getting the following error while building ParFlow’s CLM on a pretty new version of gfortran (in Amanzi, but the source is identical to ParFlow in this module, so maybe you’ve already run into this Steve).

I’m not sure where these magic constants came from, or why they aren’t valid.  Is IEEE supported in gfortran at this point?  Maybe there is a better way?  I’ve looked at ELM’s version of this, and it is pages upon pages of compiler-specific, machine specific, preprocessed coded generated by CIME, so hopefully there is something simpler than that!

Any guesses?

(from infnan.F90)

   16 |   real(r8), parameter :: inf = O'17740000000'

      |                              1

Error: BOZ literal constant at (1) is neither a data-stmt-constant nor an actual argument to INT, REAL, DBLE, or CMPLX intrinsic function [see '-fno-allow-invalid-boz']

/Users/uec/code/ats/amanzi-tpls/build-master/relwithdebinfo/clm/clm-0.1.0-source/src/utils/clm_infnan.F90:17:30:
   17 |   real(r8), parameter :: nan = O'17757777777'

      |                              1

Error: BOZ literal constant at (1) is neither a data-stmt-constant nor an actual argument to INT, REAL, DBLE, or CMPLX intrinsic function [see '-fno-allow-invalid-boz']

make[2]: *** [CMakeFiles/clm.dir/src/utils/clm_infnan.F90.o] Error 1

make[1]: *** [CMakeFiles/clm.dir/all] Error 2

make: *** [all] Error 2

--- a/pfsimulator/clm/infnan.F90
+++ b/pfsimulator/clm/infnan.F90
@@ -10,13 +10,20 @@ module infnan
 !	are being used before being set.
 !-------------------------------------------------------------------------
   use precision
-!#if (defined DOUBLE_PRECISION)
- ! real(r8), parameter :: inf = O'777600000000000000000' 
-  !real(r8), parameter :: nan = O'777677777777777777777' 
-!#else 
-  real(r8), parameter :: inf = O'17740000000'
-  real(r8), parameter :: nan = O'17757777777'
-!#endif
+! Using trick here to make inf and nan values based on octal byte
+! representations and EQUIVALNCE.
+! The more natural parameter statements caused compilation issues
+!  real(r8), parameter :: inf = O'17740000000'
+!  real(r8), parameter :: nan = O'17757777777'
+  integer :: int_infinity
+  integer :: int_nan
+  data int_infinity/O'17740000000'/
+  data int_nan/O'17757777777'/
+  real(r8) inf
+  real(r8) nan
+  EQUIVALENCE (inf,int_infinity)
+  EQUIVALENCE (nan,int_nan)
+
   integer, parameter  :: bigint = 100000000
 end module infnan
 

